File: scraping-kotlin/src/main/kotlin/utils/ScrapingUtil.kt
package utils

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.lang.StringBuilder
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.stream.Stream

class ScrapingUtil {
    companion object {
        val NAVER_SFINANCE_API_URL = "https://api.finance.naver.com/siseJson.naver"
        val params = mutableMapOf("requestType" to "1", "timeframe" to "day")

        fun scraping(code: String): String {
            val objectMapper = ObjectMapper()
            params["symbol"] = code
            params["startTime"] = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"))
            params["endTime"] = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"))

            val queryParams = params.map { (key, value) -> "$key=$value" }.joinToString("&")

            val httpClient = HttpClient.newBuilder().build()
            val request = HttpRequest.newBuilder()
                            .uri(URI.create("$NAVER_SFINANCE_API_URL?$queryParams"))
                            .build()
            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            val stringBuilder = StringBuilder("\n ### $code 주가정보 ### \n")

            if(response.statusCode() in 200..299) {
                val responseBody = response.body().replace('\'', '"')
                val menu = objectMapper.readTree(responseBody)[0].map { it.toString() }.toMutableList()
                val values = objectMapper.readTree(responseBody)[1].map { it.toString() }.toMutableList()
                menu.forEachIndexed { index, value ->  stringBuilder.append("$value : ${values[index]} \n")}
            }

            return stringBuilder.toString()
        }
    }
}

File: scraping-nodejs/utils/ScrapingUtil.js
const request = require('request');
const DateUtil = require('./DateUtil');
const FileUtil = require('./FileUtil');
module.exports = class ScrapingUtil {
    static NAVER_FINANCE_URL = 'https://api.finance.naver.com/siseJson.naver';

    scraping(code) {
        const requestOptions = {
            uri: ScrapingUtil.NAVER_FINANCE_URL,
            qs: {
                symbol : code,
                requestType : 1,
                timeframe : 'day',
                startTime : new DateUtil().getDate(),
                endTime : new DateUtil().getDate()
            }
        }
        let result = ['code : ' + code];
        request(requestOptions, function (error, response, body) {
            body = body.replace(/'/g, '"');
            const menu = JSON.parse(body)[0];
            const values = JSON.parse(body)[1];

            for(let index = 0; index < menu.length; index++) {
                result.push(' ' + menu[index] + " : " + values[index]);
            }
            new FileUtil().saveAs(result.toString() + "\n", code);
        });

    }


}
